packages/camera/camera_platform_interface/lib/src/method_channel/method_channel_camera.dart

```diff
@@ -0,0 +1,251 @@
+  }
+
+  @override
+  Stream<CameraClosingEvent> onCameraClosing(int cameraId) {
+    return _events(cameraId).whereType<CameraClosingEvent>();
+  }
+
+  @override
+  Stream<CameraErrorEvent> onCameraError(int cameraId) {
+    return _events(cameraId).whereType<CameraErrorEvent>();
+  }
+
+  @override
+  Future<XFile> takePicture(int cameraId) async {
+    String path = await _channel.invokeMethod<String>(
+      'takePicture',
+      <String, dynamic>{'cameraId': cameraId},
+    );
+    return XFile(path);
+  }
+
+  @override
+  Future<void> prepareForVideoRecording() =>
+      _channel.invokeMethod<void>('prepareForVideoRecording');
+
+  @override
+  Future<void> startVideoRecording(int cameraId) async {
+    await _channel.invokeMethod<void>(
+      'startVideoRecording',
+      <String, dynamic>{'cameraId': cameraId},
+    );
+  }
+
+  @override
+  Future<XFile> stopVideoRecording(int cameraId) async {
+    String path = await _channel.invokeMethod<String>(
+      'stopVideoRecording',
+      <String, dynamic>{'cameraId': cameraId},
+    );
+    return XFile(path);
+  }
+
+  @override
+  Future<void> pauseVideoRecording(int cameraId) => _channel.invokeMethod<void>(
+        'pauseVideoRecording',
+        <String, dynamic>{'cameraId': cameraId},
+      );
+
+  @override
+  Future<void> resumeVideoRecording(int cameraId) =>
+      _channel.invokeMethod<void>(
+        'resumeVideoRecording',
+        <String, dynamic>{'cameraId': cameraId},
+      );
+
+  @override
+  Widget buildPreview(int cameraId) {
+    return Texture(textureId: cameraId);
+  }
+
+  /// Returns the resolution preset as a String.
+  String _serializeResolutionPreset(ResolutionPreset resolutionPreset) {
+    switch (resolutionPreset) {
+      case ResolutionPreset.max:
+        return 'max';
+      case ResolutionPreset.ultraHigh:
+        return 'ultraHigh';
+      case ResolutionPreset.veryHigh:
+        return 'veryHigh';
+      case ResolutionPreset.high:
+        return 'high';
+      case ResolutionPreset.medium:
+        return 'medium';
+      case ResolutionPreset.low:
+        return 'low';
+    }
+    throw ArgumentError('Unknown ResolutionPreset value');
+  }
+
+  /// Parses a string into a corresponding CameraLensDirection.
+  @visibleForTesting
+  CameraLensDirection parseCameraLensDirection(String string) {
+    switch (string) {
+      case 'front':
+        return CameraLensDirection.front;
+      case 'back':
+        return CameraLensDirection.back;
+      case 'external':
+        return CameraLensDirection.external;
+    }
+    throw ArgumentError('Unknown CameraLensDirection value');
+  }
+
+  /// Converts messages received from the native platform into events.
+  ///
+  /// This is only exposed for test purposes. It shouldn't be used by clients of
+  /// the plugin as it may break or change at any time.
+  @visibleForTesting
+  Future<dynamic> handleMethodCall(MethodCall call, int cameraId) async {
```

Comment:
I also was not to happy with this but somehow I couldn't get it to work. Since this is a reversed method channel (i.e. sending messages from the native side to Dart) there is no use in mocking the method channel. Since the `handleMethodCall` we would like to test is the actual method callback executed when a message arrives on the method channel, when mocking the channel we would override the callback. 

On the other hand we could create a mock channel and register the `handleMethodCall` method as the callback, but this again needs it to be public and visible for testing. 

I have tried to make the `channels` array public and access it from the test directly trying to invoke methods on it and see if the streams emit expected values. But for some reasons I can get this to work at all. Breakpoints placed in the `handleMethodCall` method don't get triggered. This however would still expose the `channels` map publicly so also not really helps. 

We could refactor it into something similar like the utility methods and make it internally public and don't export it to the outside world. Something like a `MethodHandlerImpl` which gets registered during the `initialize` method. 

Would love to hear how you think about this?

Reply:
There may be a way of injecting events to the method channel through their [binaryMessenger](https://github.com/flutter/flutter/blob/master/packages/flutter_web_plugins/lib/src/plugin_registry.dart#L174), but I don't have any good example of it now. Let's leave this as @visibleForTesting for now, we can make it private later when we have a better testing strategy, thanks for digging!